### PR TITLE
feat(opensearch): implement support for opensearch acls

### LINKED ISSUE #315: Support OpenSearch ACLs
I can't find any way to manage OpenSearch ACLs. This should probably just be a quick search&replace in the Elasticsearch ACLs module, assuming they work more or less the same way.

### PR DESCRIPTION
**Pull request by mortenlj**
Fixes #315

### CODE REVIEW

**Comment by Serpentiel** [comment]
hey, @mortenlj! 👋

it is possible to manage the OpenSearch ACLs via the ElasticSearch API, have you tried to use it?

this is the way our Terraform provider manages them; I'm not sure if we would want to have duplicates of the similar endpoints

**Comment by mortenlj** [comment]
> it is possible to manage the OpenSearch ACLs via the ElasticSearch API, have you tried to use it?

No, didn't try as the Elasticsearch API isn't documented anywhere. I see now that it works.



> I'm not sure if we would want to have duplicates of the similar endpoints

For what it's worth, keeping the client code for a product you no longer offer while not having client code for the product that has replaced it seems somewhat of a strange choice. :slightly_smiling_face: 
Especially since it would seem there is some kind of compatability-layer in the API to convert from opensearch to elasticsearch, which presumably is going to be removed at some point.

File: opensearch_acls.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	path := buildPath("project", project, "service", service, "opensearch", "acl")
+	bts, err := h.client.doPutRequest(ctx, path, req)
+	if err != nil {
+		return nil, err
+	}
+
+	var r OpenSearchACLResponse
+	errR := checkAPIResponse(bts, &r)
+
+	return &r, errR
```
Comment (Serpentiel) [inline opensearch_acls.go:55]: do that instead:

```go
	var r OpenSearchACLResponse
	return &r, checkAPIResponse(bts, &r)
```

and everywhere else where applicable

File: opensearch_acls.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+// Delete subtracts ACL from already existing OpenSearch ACLs config
+func (conf *OpenSearchACLConfig) Delete(ctx context.Context, acl OpenSearchACL) *OpenSearchACLConfig {
```
Comment (Serpentiel) [inline opensearch_acls.go:73]: this is quite complex, try something like that instead:

```go
// Delete removes the specified ACL from the existing OpenSearch ACLs config.
func (conf *OpenSearchACLConfig) Delete(ctx context.Context, acl OpenSearchACL) *OpenSearchACLConfig {
	newACLs := []OpenSearchACL{}  // Create a new slice to hold the updated list of ACLs.

	// Iterate over each existing ACL entry.
	for _, existingAcl := range conf.ACLs {
		// If the ACL usernames match, we'll potentially modify the rules.
		if acl.Username == existingAcl.Username {
			newRules := []Rule{}  // Create a new slice to hold the updated list of rules.

			// Check each existing rule against the rules in the ACL to be deleted.
			for _, existingRule := range existingAcl.Rules {
				match := false  // Flag to track if the existing rule matches any rule in the ACL to be deleted.
				for _, ruleToDelete := range acl.Rules {
					if existingRule.Permission == ruleToDelete.Permission && existingRule.Index == ruleToDelete.Index {
						match = true  // The existing rule matches a rule in the ACL to be deleted.
						break
					}
				}
				// If the existing rule doesn't match any rule in the ACL to be deleted, add it to the new list.
				if !match {
					newRules = append(newRules, existingRule)
				}
			}

			// If there are remaining rules after deletion, add the modified ACL to the new list.
			if len(newRules) > 0 {
				existingAcl.Rules = newRules
				newACLs = append(newACLs, existingAcl)
			}
		} else {
			// If the usernames don't match, directly add the existing ACL to the new list.
			newACLs = append(newACLs, existingAcl)
		}
	}

	// Replace the original list of ACLs with the updated list.
	conf.ACLs = newACLs
	return conf
}
```

File: opensearch_acls.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+				}
+			}
+		}
+	}
+
+	return conf
+}
+
+// Add appends new ACL to already existing OpenSearch ACLs config
+func (conf *OpenSearchACLConfig) Add(acl OpenSearchACL) *OpenSearchACLConfig {
```
Comment (Serpentiel) [inline opensearch_acls.go:96]: this is also quite complex, how about that:

```go
// Add appends new ACL to the existing OpenSearch ACLs config.
func (conf *OpenSearchACLConfig) Add(acl OpenSearchACL) *OpenSearchACLConfig {
	var userIndex int
	userExists := false

	// Iterate over the existing ACLs to identify duplicates and determine user existence.
	for p, existingAcl := range conf.ACLs {
		if acl.Username == existingAcl.Username {
			userExists = true
			userIndex = p

			// Filter out any rules in the ACL to add that already exist for the user.
			remainingRules := []Rule{}
			for _, rule := range acl.Rules {
				exists := false
				for _, existingRule := range existingAcl.Rules {
					if rule.Permission == existingRule.Permission && rule.Index == existingRule.Index {
						exists = true
						break
					}
				}
				if !exists {
					remainingRules = append(remainingRules, rule)
				}
			}
			acl.Rules = remainingRules
		}
	}

	// If no rules remain for the user, return the existing configuration.
	if len(acl.Rules) == 0 {
		return conf
	}

	// Add the new or updated ACL to the config.
	if userExists {
		conf.ACLs[userIndex].Rules = append(conf.ACLs[userIndex].Rules, acl.Rules...)
	} else {
		conf.ACLs = append(conf.ACLs, acl)
	}

	return conf
}
```

**Review by Serpentiel — CHANGES_REQUESTED** [review]

**Comment by Serpentiel** [comment]
also, I suggest that you update ElasticSearch part of the client and mark the methods as deprecated

File: opensearch_acls.go (hunk 1 above)
Comment (mortenlj) [reply opensearch_acls.go]: I only did the ones in `opensearch_acls.go` and `elasticsearch_acls.go`. Felt making this change everywhere belongs in a separate PR.

File: opensearch_acls.go (hunk 3 above)
Comment (mortenlj) [reply opensearch_acls.go:110]: Done

File: opensearch_acls.go (hunk 1 above)
Comment (Serpentiel) [reply opensearch_acls.go]: I meant within the scope of this PR :D

**Comment by Serpentiel** [comment]
hmm, looks like `commitlint` has failed on this PR because of your 2nd commit; could you perhaps squash the commits in the single, original one? thank you!

**Comment by mortenlj** [comment]
> hmm, looks like `commitlint` has failed on this PR because of your 2nd commit; could you perhaps squash the commits in the single, original one? thank you!

Tried doing that, only to now realize I've picked up a random dependabot commit on the way. Not sure how to undo that, my git fu isn't that strong ...

**Comment by Serpentiel** [comment]
hey, @mortenlj! 👋

I could fix it for you if you [allow edits of this PR for maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) in the PR

**Comment by mortenlj** [comment]
> I could fix it for you if you [allow edits of this PR for maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) in the PR

I couldn't find that option, but got some help with the git commands and managed to fix it I believe.

**Review by Serpentiel — APPROVED** [review]
lgtm

**Pull request merged**